fix: make fips compliant by replacing md5 with sha256 in postgres and usedforsecurity=False for python md5

actionjax opened the pull request:
<!---
Please write the PR title following the conventions at https://www.conventionalcommits.org/en/v1.0.0/
Example:
-->
fix(superset): allows superset to run on fips compliance machines

### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
added usedforsecurity=False for md5(..) methods and replaced postgres md5(...) with sha256(..)

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
<!--- Skip this if not applicable -->

### TESTING INSTRUCTIONS
<!--- Required! What steps can be taken to manually verify the changes? -->
Run a fips complaint Linux VM (we used Rocky8) but any Linux with fips enabled should work. 

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [x] Has associated issue: #20904 
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

actionjax committed:
fix: make fips compliant by adding useforsecurity=False flag on md5 hashes

craig-rueda commented on superset/key_value/utils.py:
```diff
@@ -61,9 +61,9 @@ def decode_permalink_id(key: str, salt: str) -> int:
-    md5_obj = md5()
-    md5_obj.update(seed.encode("utf-8"))
-    return UUID(md5_obj.hexdigest())
+    rd = random.Random()
```
Won't this break existing namespaces? It seems that this logic was intended to keep UUIDs deterministic based on some input

actionjax replied:
Yes, at first I thought the uuid was only used in cache, looks like that hash is stored in the db. To address your concern about deterministic I've replaced random with sha256. Will look to see how we can make this backwards compatible.

villebro replied:
I think you would need to make a migration that updates all existing entries from the MD5 based UUIDs to using SHA256 as the base. I think it should be doable, but we do need to be very careful with this. Ping @michael-s-molina , do you have any concerns here?

craig-rueda commented on superset/utils/hashing.py:
```diff
@@ -21,7 +21,7 @@
 
 def md5_sha_from_str(val: str) -> str:
-    return hashlib.md5(val.encode("utf-8")).hexdigest()
+    return hashlib.sha256(val.encode("utf-8")).hexdigest()
```
Same comment as above

craig-rueda reviewed (commented):
Looks like there's a few potential issues with this. Pls provide some tests that validate backwards compatibility.

ccravens commented:
WIth the recent release of 3.0.0 was hoping to see this as a lot of orgs run on FIPS compliant environments and this prevents superset from operating in those environments.

rusackas commented:
Seems there are some loose threads here... is anyone planning to pick this back up, or should we close it out?

villebro commented:
@actionjax it should be easy to make this configurable, thus keeping backwards compatibility. Let me know if you need help implementing it.

actionjax commented:
> @actionjax it should be easy to make this configurable, thus keeping backwards compatibility. Let me know if you need help implementing it.

Hey @villebro, unfortunately I do not have the cycles to look and implement this currently. I would recommend someone else taking over.

rusackas commented:
Converting this to draft mode, rather than closing this. In due time, if nobody picks it up, we might close it out, but I remain hopeful. I'll close and re-open it just to jump-start CI and see how that plays out.

villebro commented:
Ping @ccravens if you're interested in adding optional support for FIPS compliance

digitalkristurner commented:
I would be interested in working on this. I think Superset is a great project and would be really helpful in other projects I'm working on.

rusackas commented:
@digitalkristurner fantastic! We'd love to see that.

digitalkristurner commented:
So I've started and found all the references to MD5 - everything still builds so that's a start. I'm trying to understand the unit_tests/thumbnails/test_digest.py (and the multiple other tests) though. These have hardcoded MD5s so will need to convert these but have no idea what they've been generated off.

digitalkristurner commented:
Hey all, changed all the references from MD5 to SHA256 and then checked and corrected for all the possible overflows within the databases. I've ran the unit tests and they're all passing now. Just need to try and figure out the integration tests - any steer or guidance on that would be much appreciated.

Quick note postgres only added support for SHA256 in Postgres 11 - not sure how much of a problem that'll be?

actionjax commented:
> Hey all, changed all the references from MD5 to SHA256 and then checked and corrected for all the possible overflows within the databases. I've ran the unit tests and they're all passing now. Just need to try and figure out the integration tests - any steer or guidance on that would be much appreciated.
> 
> Quick note postgres only added support for SHA256 in Postgres 11 - not sure how much of a problem that'll be?

Hey @digitalkristurner, sorry for the late reply, one of my coworkers pointed out that since the md5 hash isn't really used for security purposes in superset but rather just to create a UUID / hash we should be able to get away with just changing the code to have a security flag set to false

```
md5(..., usedforsecurity=False)
```

this should pass fips security requirements, if the superset team agrees, I think this is the path of least resistance.

digitalkristurner commented:
> > Hey all, changed all the references from MD5 to SHA256 and then checked and corrected for all the possible overflows within the databases. I've ran the unit tests and they're all passing now. Just need to try and figure out the integration tests - any steer or guidance on that would be much appreciated.
> 
> > 
> 
> > Quick note postgres only added support for SHA256 in Postgres 11 - not sure how much of a problem that'll be?
> 
> 
> 
> Hey @digitalkristurner, sorry for the late reply, one of my coworkers pointed out that since the md5 hash isn't really used for security purposes in superset but rather just to create a UUID / hash we should be able to get away with just changing the code to have a security flag set to false
> 
> 
> 
> ```
> 
> md5(..., usedforsecurity=False)
> 
> ```
> 
> 
> 
> this should pass fips security requirements, if the superset team agrees, I think this is the path of least resistance.

Will that work for the MD5 reference in the PGDialect uuid_by_dialect call? I'm not sure how Postgres deals with FIPS.

andrewzah commented:
> the md5 hash isn't really used for security purposes

Was this tested on a fips-enabled system? Attempting to use md5 at all should error, giving something like this

```
$ openssl md5 /dev/null

Error setting digest md5
140647163811744:error:060800A3:digital 
```

digitalkristurner commented:
I'll build try building the container and deploying it in our dev environment.

actionjax commented:
> > the md5 hash isn't really used for security purposes
> 
> Was this tested on a fips-enabled system? Attempting to use md5 at all should error, giving something like this
> 
> ```
> $ openssl md5 /dev/null
> 
> Error setting digest md5
> 140647163811744:error:060800A3:digital 
> ```

ah yea, I overlooked the postgres part. I don't think there's a workaround for that unlike python. I think it will complain so you can continue the work you've already done.

digitalkristurner commented:
I tried your method and it seemed to work - it ran within a FIPS environment for me. Might just need to change the two strings within the DB but otherwise everything else works as usual.

digitalkristurner commented:
I'm not sure what I'm supposed to do now to share the code changes I made (considering this isn't my pull request).

actionjax commented:
> I'm not sure what I'm supposed to do now to share the code changes I made (considering this isn't my pull request).

I updated the code based on your findings @digitalkristurner, does this look correct? I don't mind you forking the code and submitting the same pull request  for credit

digitalkristurner commented:
Yeah that looks right and then `PGDialect: "uuid_in(sha256(random()::text || clock_timestamp()::text)::cstring)"` in superset/migrations/shared.

actionjax committed:
fix: make fips compliant by replacing md5 with sha256 in PGDialect

villebro commented:
@actionjax could you update the description to reflect the current state of the PR? Also, I'm not sure we can universally declare `usedforsecurity=False`, as different security teams may have a different stance on what constitutes security usage:

![image](https://github.com/apache/superset/assets/33317356/cd8d58e4-5c4c-4cd9-9fd1-c763505aa97d)

Source: https://docs.python.org/3/library/hashlib.html#constructors

I'd be curious to hear other security experts chime in here. Also, I think it'd be a good idea to discuss this in the next town hall before proceeding with this change, and maybe even attaching a SIP to it. Ping @michael-s-molina

digitalkristurner commented:
It's not a lot of work changing it completely to SHA256 - I got that building and the unit tests worked (just had to change the set variables). I never got the integration tests to run properly so can't confirm that.

Yuval-Moshe commented:
Hi @actionjax @digitalkristurner , is there any intention to pick this one up? we also need superset to be FIPS complaint as it seems many others are, and it looks like the majority of the job was already done in this PR.

Thanks!

fred-hartman commented:
My company also needs to run Superset in FIPS mode.  Does anyone know the true state of this PR?  Or if there is another discussion/plan for FIPS support?

The files changed https://github.com/apache/superset/pull/21014/files do not change all md5 use to sha256 as [digitalkristurner](https://github.com/digitalkristurner) implied and from my naive reading of the code and [craig-rueda](https://github.com/craig-rueda) statement, there could be some compatibility issues with just changing each of the lines of code in this PR.

villebro commented:
@fred-hartman my recommendation going forward is as follows:
1. We make FIPS compliance optional by introducing configurable flags/hooks for calculating hashes that default to the current implementation (MD5), but would support replacing those with a FIPS compliant variant (SHA256).
1. In a forthcoming major version we make a breaking change, where we start defaulting to FIPS compliance being enabled, but support running in non-compliant mode. Then all current deployments would need to explicitly configure their deployments as non-FIPS compliant, ensuring they continue working as expected, but new deployments would be FIPS compliant by default.

This would require a SIP, as this is a pretty significant change. Also note that I don't believe any of the core contributors are working on this, so it would need to be a community driven effort. But I'm happy to help push it forward if someone can drive the actual SIP and implementation work.

fred-hartman commented:
Thanks for the details.
> @fred-hartman my recommendation going forward is as follows:
> 
> 1. We make FIPS compliance optional by introducing configurable flags/hooks for calculating hashes that default to the current implementation (MD5), but would support replacing those with a FIPS compliant variant (SHA256).
> 2. In a forthcoming major version we make a breaking change, where we start defaulting to FIPS compliance being enabled, but support running in non-compliant mode. Then all current deployments would need to explicitly configure their deployments as non-FIPS compliant, ensuring they continue working as expected, but new deployments would be FIPS compliant by default.
> 

That makes sense as a rollout plan. HITRUST certification next year will be adopting many FEDRAMP requirements, including the requirement of running in FIPS mode, so I expect demand will grow.

> This would require a SIP, as this is a pretty significant change. Also note that I don't believe any of the core contributors are working on this, so it would need to be a community driven effort. But I'm happy to help push it forward if someone can drive the actual SIP and implementation work.

SHA254 hashes are 64 chars and MD5 are 32. Do you know if MD5 values are persisted in the DB? That would increase the scope of migration significantly. A quick look at the DB tables don't show any VARCHAR(32) columns, but a few with VARCHAR(50), which I haven't tracked to how they are filled.

I'll have to escalate internally to see if I can get a python resource to be hands on this or if we look for an outside contractor to craft a PR.

Yuval-Moshe commented:
Ok, thank you. Our water pump is not working either, also happened twice in
the past couple of days (we called and notified it as well).
Can the mechanic take a look and fix it tomorrow as well?
We currently don't have running water and we just filled the water tank
yesterday.

On Tue, 10 Dec 2024, 11:37 Fred Hartman, ***@***.***> wrote:

> Thanks for the details.
>
> @fred-hartman <https://github.com/fred-hartman> my recommendation going
> forward is as follows:
>
>    1. We make FIPS compliance optional by introducing configurable
>    flags/hooks for calculating hashes that default to the current
>    implementation (MD5), but would support replacing those with a FIPS
>    compliant variant (SHA256).
>    2. In a forthcoming major version we make a breaking change, where we
>    start defaulting to FIPS compliance being enabled, but support running in
>    non-compliant mode. Then all current deployments would need to explicitly
>    configure their deployments as non-FIPS compliant, ensuring they continue
>    working as expected, but new deployments would be FIPS compliant by default.
>
> That makes sense as a rollout plan. HITRUST certification next year will
> be adopting many FEDRAMP requirements, including the requirement of running
> in FIPS mode, so I expect demand will grow.
>
> This would require a SIP, as this is a pretty significant change. Also
> note that I don't believe any of the core contributors are working on this,
> so it would need to be a community driven effort. But I'm happy to help
> push it forward if someone can drive the actual SIP and implementation work.
>
> SHA254 hashes are 64 chars and MD5 are 32. Do you know if MD5 values are
> persisted in the DB? That would increase the scope of migration
> significantly. A quick look at the DB tables don't show any VARCHAR(32)
> columns, but a few with VARCHAR(50), which I haven't tracked to how they
> are filled.
>
> I'll have to escalate internally to see if I can get a python resource to
> be hands on this or if we look for an outside contractor to craft a PR.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/superset/pull/21014#issuecomment-2529683581>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AMZXEBVL2O4FOYOYTWZWEJD2EYLTNAVCNFSM6AAAAABPWCRADOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKMRZGY4DGNJYGE>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

villebro commented:
> SHA254 hashes are 64 chars and MD5 are 32. Do you know if MD5 values are persisted in the DB? That would increase the scope of migration significantly. A quick look at the DB tables don't show any VARCHAR(32) columns, but a few with VARCHAR(50), which I haven't tracked to how they are filled.

In Superset, it's primarily only used to calculate a deterministic short representation of a complex object, especially during caching, and for this purpose any other hashing algorithm can also be used. Again, the main concern is backwards compatibility - when you switch over from MD5 to SHA256, you will get 100 % cache misses, as the old values will be cached with MD5, but you will try to retrieve them using SHA256.

Btw, it's probably also important to double check Superset's dependencies, as they may also be using MD5. Not sure if that's already been audited, but I'd almost be surprised if you don't run into some issues in the deps.

rusackas commented:
Shall we close this one out? It's been inactive since December, so I'm not sure if it's likely to move foward.
